amulet/sentry.py

```diff
@@ -307,28 +344,149 @@ def wait(self, timeout=300):
+            list is given, then there must be a one-to-one match between the
+            messages and the units, though the order doesn't matter.
+        :param int timeout: Number of seconds to wait before timing-out.
+
+        Examples::
+
+            # wait for all units to report "ready"
+            t.wait_for_messages({'ubuntu': 'ready'})
+
+            # wait for all units to report something like "ready"
+            t.wait_for_messages({'ubuntu': re.compile('r..dy')})
+
+            # wait for at least one unit to report "ready"
+            t.wait_for_messages({'ubuntu': {'ready'})
+
+            # wait for all units to report either "ready" or "ok"
+            t.wait_for_messages({'ubuntu': re.compile('ready|ok')})
+
+            # wait for at least one unit to report "ready" and at least one
+            # unit to report "ok"
+            t.wait_for_messages({'ubuntu': {'ready', 'ok'}})
+
+            # wait for one unit to report "ready" and the other to report "ok"
+            # (must be exactly two units)
+            t.wait_for_messages({'ubuntu': ['ready', 'ok']})
+        """
+        def get_messages(service, status):
+            messages = []
+            for unit in status[service].values():
+                if not unit['workload-status']:
+                    raise helpers.UnsupportedError()
+                messages.append(unit['workload-status'].get('message', ''))
+            return messages
+
+        matcher = StatusMessageMatcher()
+        for i in helpers.timeout_gen(timeout):
+            status = self.get_status()
+            for service, expected in messages.items():
+                actual = get_messages(service, status)
+                if not matcher.check(expected, actual):
+                    break
+            else:
+                return
 
     def _sync(self):
         pass
 
 
+class StatusMessageMatcher(object):
+    def check(self, expected, actual):
+        if isinstance(expected, (list, tuple)):
+            return self.check_list(expected, actual)
+        elif isinstance(expected, set):
+            return self.check_set(expected, actual)
+        else:
+            return self.check_messages(expected, actual)
+
+    def check_messages(self, expected, actual):
+        if not actual:
+            return False
+        for a in actual:
+            if not self.check_message(expected, a):
+                return False
+        return True
+
+    def check_set(self, expected, actual):
+        if not actual:
+            return False
+        for e in expected:
+            for a in actual:
+                if self.check_message(e, a):
+                    break  # match, go to next expected
+            else:
+                return False  # no match for this expected (all must match)
+        return True
+
+    def check_list(self, expected, actual):
+        """
+        Lists must match one-to-one.
+
+        pathlogocial: ['ready', 'ready', 'ok'] vs ['ready', 'ok', 'ok']
```

Comment:
(sp) pathological. Also please clarify what are these docstring examples are intended to show.